imports/client/components/FancyEditor.tsx

```diff
@@ -130,17 +144,65 @@ const EditableMentionRenderer = ({
   );
 };
 
+const Image = styled.img`
+  max-height: 20px;
+`;
+const ChatImageRenderer = ({
+  attributes,
+  element,
+}: ElementRendererProps<ImageElement>) => {
+  return (
+    <Image
+      {...attributes}
+      src={element.url}
+      title={element.status === "error" ? "Image upload failed" : ""}
+    />
+  );
+};
+
 // Composed, layered reassignment is how Slate plugins are designed to work.
-const withMentions = (editor: Editor) => {
-  const { isInline, isVoid, markableVoid } = editor;
+const withMentionsAndImages = (
+  editor: Editor,
+  uploadImageFile: (file: File) => void,
+) => {
+  const { isInline, isVoid, markableVoid, insertData } = editor;
   editor.isInline = (element) => {
-    return element.type === "mention" ? true : isInline(element);
+    switch (element.type) {
+      case "mention":
+      case "image":
+        return true;
+      default:
+        return isInline(element);
+    }
   };
   editor.isVoid = (element) => {
-    return element.type === "mention" ? true : isVoid(element);
+    switch (element.type) {
+      case "mention":
+      case "image":
+        return true;
+      default:
+        return isVoid(element);
+    }
   };
   editor.markableVoid = (element) => {
-    return element.type === "mention" || markableVoid(element);
+    switch (element.type) {
+      case "mention":
+        return true;
+      case "image":
+        return false;
+      default:
+        return markableVoid(element);
+    }
+  };
+  editor.insertData = (data: DataTransfer) => {
+    if (!data) return;
+    for (const file of Array.from(data.files)) {
+      if (file.type.startsWith("image/")) {
+        uploadImageFile(file);
+        return;
+      }
```

Comment:
I think this has some slightly weird interactions with the fact that `createChatImageUpload` doesn't do any validation of the submitted mime type.

Setting `accept="image/*"` on the file input means that my browser defaults to image-only uploads, but I can override that to upload any file type. If I do that and upload, say, a `text/plain` file, it gets uploaded to S3 and then disappears - I think because this check short-circuits.

My instinct is that we should check and refuse to upload files that aren't mime type `image/*` - probably by checking both in `createChatImageUpload` and `uploadImageFile`